### ISSUE TITLE
aws_sns: Add support for $or operator in subscription filter policies

### ISSUE THREAD

**Issue by ross-bragg-sonarsource** [aws/aws-cdk#30524]
### Describe the feature

https://docs.aws.amazon.com/sns/latest/dg/and-or-logic.html#or-operator

I would like to be able to use the "$or" operator for SNS subscription filters to filter out messages based on multiple combinations of message attributes.

### Use Case

I would like to create the following filter policy on an SNS subscription:

```json
{
  "$or": [
    {
      "attribute_1": [
        "value_1",
        "value_2"
      ]
    },
    {
      "attribute_1": [
        "value_3"
      ],
      "attribute_2": [
        "value_4"
      ]
    }
  ]
}
```

However, the `filter_policy` on an SNS subscription is a `Mapping[str, SubscriptionFilter]`, which doesn't account for the "$or" operator which effectively takes a list of filter policies. https://docs.aws.amazon.com/cdk/api/v2/python/aws_cdk.aws_sns/Subscription.html

I can get my desired result, but I have to set the filter policy directly:
```python
subscription = self._topic.add_subscription(
    aws_sns_subscriptions.SqsSubscription(
        self._queue,
    )
)

subscription.node.default_child.filter_policy = filter_policy
```
where filter_policy is the dictionary at the beginning of the section.

I was unable to find if this was already supported in the documentation.

### Acknowledgements

- [ ] I may be able to implement this feature request
- [ ] This feature might incur a breaking change

### CDK version used

2.141.0

### Environment details (OS name and version, etc.)

MacOS, Python 3.9.6

**Comment by nmussy** [comment]
I don't think we can forego a typing change to fix this issue, whatever we do will still break the `{ [attribute: string]: SubscriptionFilter }` typing.

The least invasive change would be to add a static `SubscriptionFilter.or` method to sugarcoat the operator, but it might be worth also adding a static `SubscriptionFilter.and` method for the sake of consistency? I'm not sure this is the way to go, it does add some unecessary synthax and will require to check that there are no duplicate keys in the `and(...filters: { [attribute: string]: SubscriptionFilter }[])` parameters.

```ts
type SubscriptionFilterMap = { [attribute: string]: SubscriptionFilter | SubscriptionFilterMap[] };

export interface SubscriptionOptions {
  // ...
  readonly filterPolicy? : SubscriptionFilterMap;
}

export class SubscriptionFilter {
  public static and(...filters: { [attribute: string]: SubscriptionFilter }[]) {
    // TODO check conflicting filter keys
    return Object.assign({}, ...filters);
  }

  public static or(...filters: { [attribute: string]: SubscriptionFilter }[]) {
    return { $or: filters };
  }

  // ...
}

```

Using the `.or` operator:
```ts
new sns.Subscription(stack, 'Subscription', {
  endpoint: 'endpoint',
  filterPolicy: SubscriptionFilter.or(
    {
      color: sns.SubscriptionFilter.stringFilter({ allowlist: ['red', 'green'] }),
      price: sns.SubscriptionFilter.numericFilter({ allowlist: [100] }),
    },
    {
      color: sns.SubscriptionFilter.stringFilter({ allowlist: ['green'] }),
      price: sns.SubscriptionFilter.numericFilter({ allowlist: [500] }),
    },
    { color: sns.SubscriptionFilter.stringFilter({ allowlist: ['purple'] }) },
  ),
  protocol: sns.SubscriptionProtocol.LAMBDA,
  topic,
});
```

And also using the `.and` operator:
```ts
new sns.Subscription(stack, 'Subscription', {
  endpoint: 'endpoint',
  filterPolicy: SubscriptionFilter.or(
    SubscriptionFilter.and(
      { color: sns.SubscriptionFilter.stringFilter({ allowlist: ['red', 'green'] }) },
      { price: sns.SubscriptionFilter.numericFilter({ allowlist: [100] }) },
    ),
    SubscriptionFilter.and(
      { color: sns.SubscriptionFilter.stringFilter({ allowlist: ['green'] }) },
      { price: sns.SubscriptionFilter.numericFilter({ allowlist: [500] }) },
    ),
    { color: sns.SubscriptionFilter.stringFilter({ allowlist: ['purple'] }) },
  ),
  protocol: sns.SubscriptionProtocol.LAMBDA,
  topic,
});
```

Let me know if you can come up with a cleaner solution!

**Comment by omrib15** [comment]
any update on this? I really need to be able to us $or in my filters

**Comment by albertodiazdorado** [comment]
It would be lovely if we could do `$or` filters using CDK, without resorting to `CfnSubscription`

**Comment by orekav** [comment]
+1

**Comment by orekav** [comment]
My workaround

```typescript
class CustomSubscriptionFilter extends sns.SubscriptionFilter {
  public static customFilter(conditions: Array<any>) {
    return new sns.SubscriptionFilter(conditions)
  }
}

const deadLetterQueue = new sqs.Queue(this, "DeadLetterQueue", {
  enforceSSL: true,
  retentionPeriod: cdk.Duration.days(14),
});
const queue = new sqs.Queue(this, "Queue", {
  enforceSSL: true,
  retentionPeriod: cdk.Duration.days(14),
  visibilityTimeout: cdk.Duration.minutes(5),
  deadLetterQueue: {
    queue: deadLetterQueue,
    maxReceiveCount: 5,
  },
});

// Add subscription to the topic
props.requestDeletionTopic.addSubscription(
  new snsSubscriptions.SqsSubscription(queue, {
    rawMessageDelivery: true,
    filterPolicyWithMessageBody: {
      // LINK - https://github.com/aws/aws-cdk/issues/30524
      $or: sns.FilterOrPolicy.filter(
        CustomSubscriptionFilter.customFilter([
          {
            version: [1],
            isPerson: [true],
          },
          {
            version: [2],
            type: ['person', 'animal'],
            'entity.name': ['Test'],
          },
        ]),
      ),
    },
  }),
)
```

**Comment by henhal** [comment]
Hi @orekav 
This workaround relies on the filter policy being on a message body level rather than message attributes, right?

I still cannot see any way to add an `$or` policy on a message attribute level without resorting to L1 constructs...

> My workaround
> 
> class CustomSubscriptionFilter extends sns.SubscriptionFilter {
>   public static customFilter(conditions: Array<any>) {
>     return new sns.SubscriptionFilter(conditions)
>   }
> }
> 
> const deadLetterQueue = new sqs.Queue(this, "DeadLetterQueue", {
>   enforceSSL: true,
>   retentionPeriod: cdk.Duration.days(14),
> });
> const queue = new sqs.Queue(this, "Queue", {
>   enforceSSL: true,
>   retentionPeriod: cdk.Duration.days(14),
>   visibilityTimeout: cdk.Duration.minutes(5),
>   deadLetterQueue: {
>     queue: deadLetterQueue,
>     maxReceiveCount: 5,
>   },
> });
> 
> // Add subscription to the topic
> props.requestDeletionTopic.addSubscription(
>   new snsSubscriptions.SqsSubscription(queue, {
>     rawMessageDelivery: true,
>     filterPolicyWithMessageBody: {
>       // LINK - https://github.com/aws/aws-cdk/issues/30524
>       $or: sns.FilterOrPolicy.filter(
>         CustomSubscriptionFilter.customFilter([
>           {
>             version: [1],
>             isPerson: [true],
>           },
>           {
>             version: [2],
>             type: ['person', 'animal'],
>             'entity.name': ['Test'],
>           },
>         ]),
>       ),
>     },
>   }),
> )
> cc [@albertodiazdorado](https://github.com/albertodiazdorado)

**Comment by orekav** [comment]
> Hi [@orekav](https://github.com/orekav) This workaround relies on the filter policy being on a message body level rather than message attributes, right?
> 
> I still cannot see any way to add an `$or` policy on a message attribute level without resorting to L1 constructs...
> 
> > My workaround
> > class CustomSubscriptionFilter extends sns.SubscriptionFilter {
> > public static customFilter(conditions: Array) {
> > return new sns.SubscriptionFilter(conditions)
> > }
> > }
> > const deadLetterQueue = new sqs.Queue(this, "DeadLetterQueue", {
> > enforceSSL: true,
> > retentionPeriod: cdk.Duration.days(14),
> > });
> > const queue = new sqs.Queue(this, "Queue", {
> > enforceSSL: true,
> > retentionPeriod: cdk.Duration.days(14),
> > visibilityTimeout: cdk.Duration.minutes(5),
> > deadLetterQueue: {
> > queue: deadLetterQueue,
> > maxReceiveCount: 5,
> > },
> > });
> > // Add subscription to the topic
> > props.requestDeletionTopic.addSubscription(
> > new snsSubscriptions.SqsSubscription(queue, {
> > rawMessageDelivery: true,
> > filterPolicyWithMessageBody: {
> > // LINK - [#30524](https://github.com/aws/aws-cdk/issues/30524)
> > $or: sns.FilterOrPolicy.filter(
> > CustomSubscriptionFilter.customFilter([
> > {
> > version: [1],
> > isPerson: [true],
> > },
> > {
> > version: [2],
> > type: ['person', 'animal'],
> > 'entity.name': ['Test'],
> > },
> > ]),
> > ),
> > },
> > }),
> > )
> > cc [@albertodiazdorado](https://github.com/albertodiazdorado)

Exactly, it filters based on body. So far I am still using this approach and I works just find.
I haven't spent time on checking how to achieve this in CDK because it also has some math to do to avoid passing the conditions limit
